x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java

```diff
@@ -9,134 +9,189 @@
+            return CharArrays.constantTimeEquals(data.getChars(), hash);
         }
-    },
 
-    BCRYPT4() {
         @Override
-        public char[] hash(SecureString text) {
-            String salt = BCrypt.gensalt(4);
-            return BCrypt.hashpw(text, salt).toCharArray();
+        public String getAlgorithm() {
+            return "noop";
         }
 
         @Override
-        public boolean verify(SecureString text, char[] hash) {
-            String hashStr = new String(hash);
-            if (!hashStr.startsWith(BCRYPT_PREFIX)) {
-                return false;
-            }
-            return BCrypt.checkpw(text, hashStr);
+        public int getCost() {
+            throw new UnsupportedOperationException("Algorithm is not associated with a cost factor");
+        }
+    }
+
+    class BcryptHasher implements Hasher {
+        private int cost;
+        private static final String BCRYPT_PREFIX = "$2a$";
+
+        public BcryptHasher(int cost) {
+            this.cost = cost;
         }
-    },
 
-    BCRYPT5() {
         @Override
-        public char[] hash(SecureString text) {
-            String salt = BCrypt.gensalt(5);
-            return BCrypt.hashpw(text, salt).toCharArray();
+        public char[] hash(SecureString data) {
+            String salt = BCrypt.gensalt(this.cost);
+            return BCrypt.hashpw(data, salt).toCharArray();
         }
 
         @Override
-        public boolean verify(SecureString text, char[] hash) {
+        public boolean verify(SecureString data, char[] hash) {
             String hashStr = new String(hash);
             if (!hashStr.startsWith(BCRYPT_PREFIX)) {
                 return false;
             }
-            return BCrypt.checkpw(text, hashStr);
+            return BCrypt.checkpw(data, hashStr);
         }
-    },
 
-    BCRYPT6() {
         @Override
-        public char[] hash(SecureString text) {
-            String salt = BCrypt.gensalt(6);
-            return BCrypt.hashpw(text, salt).toCharArray();
+        public String getAlgorithm() {
+            return "bcrypt";
         }
 
         @Override
-        public boolean verify(SecureString text, char[] hash) {
-            String hashStr = new String(hash);
-            if (!hashStr.startsWith(BCRYPT_PREFIX)) {
-                return false;
-            }
-            return BCrypt.checkpw(text, hashStr);
+        public int getCost(){ return this.cost; }
+    }
+
+    class PBKDF2Hasher implements Hasher {
+        private int cost;
+        private static final String PBKDF2_PREFIX = "{PBKDF2}";
+
+        public PBKDF2Hasher(int cost) {
+            this.cost = cost;
         }
-    },
 
-    BCRYPT7() {
         @Override
-        public char[] hash(SecureString text) {
-            String salt = BCrypt.gensalt(7);
-            return BCrypt.hashpw(text, salt).toCharArray();
+        public char[] hash(SecureString data) {
+            try {
+                StringBuilder result = new StringBuilder();
```

Comment:
rather than use a string builder, can we use a CharBuffer? This avoids the creation of the string of the hash that goes into the string table and instead we just keep it in a char[]